sub-packages/bionemo-esm2/src/bionemo/esm2/testing/compare.py

```diff
@@ -0,0 +1,99 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import gc
+from pathlib import Path
+
+import torch
+from megatron.core.transformer.module import Float16Module
+from transformers import AutoModelForMaskedLM
+
+from bionemo.core.utils.dtypes import PrecisionTypes, get_autocast_dtype
+from bionemo.esm2.data.tokenizer import get_tokenizer
+from bionemo.esm2.model.model import ESM2Config
+
+
+def assert_model_equivalence(
+    ckpt_path: Path | str,
+    model_tag: str,
+    precision: PrecisionTypes = "fp32",
+    rtol: float | None = None,
+    atol: float | None = None,
+) -> None:
+    """Testing utility to compare the outputs of a NeMo2 checkpoint to the original HuggingFace model weights.
+
+    Compares the cosine similarity of the logit and hidden state outputs of a NeMo2 model checkpoint to the outputs of
+    the corresponding HuggingFace model.
+
+    Args:
+        ckpt_path: A path to a NeMo2 checkpoint for an ESM-2 model.
+        model_tag: The HuggingFace model tag for the model to compare against.
+        precision: The precision type to use for the comparison. Defaults to "fp32".
+        rtol: The relative tolerance to use for the comparison. Defaults to None, which chooses the tolerance based on
+            the precision.
+        atol: The absolute tolerance to use for the comparison. Defaults to None, which chooses the tolerance based on
+            the precision.
+    """
+    tokenizer = get_tokenizer()
+
+    test_proteins = [
+        "MKTVRQERLKSIVRILERSKEPVSGAQLAEELSVSRQVIVQDIAYLRSLGYNIVATPRGYVLA",
+        "MKTVRQERLKSI<mask>RILERSKEPVSGAQLAEELS<mask>SRQVIVQDIAYLRSLGYN<mask>VATPRGYVLAGG",
+    ]
+    tokens = tokenizer(test_proteins, return_tensors="pt", padding=True, truncation=True).to("cuda")
+    input_ids = tokens["input_ids"]
+    attention_mask = tokens["attention_mask"]
+
+    dtype = get_autocast_dtype(precision)
+    nemo_config = ESM2Config(
+        initial_ckpt_path=str(ckpt_path),
+        include_embeddings=True,
+        include_hiddens=True,
+        params_dtype=dtype,
+        pipeline_dtype=dtype,
+        autocast_dtype=dtype,
+        bf16=dtype is torch.bfloat16,
+        fp16=dtype is torch.float16,
+    )
+
+    nemo_model = nemo_config.configure_model(tokenizer).to("cuda").eval()
+
+    if dtype is torch.float16 or dtype is torch.bfloat16:
+        nemo_model = Float16Module(nemo_config, nemo_model)
+
+    nemo_output = nemo_model(input_ids, attention_mask)
```

Comment:
@jstjohn, could we use `Trainer` to call forward here?

Reply:
You can try `trainer.predict(...)` which calls `model.predict_step`. If you grep predict_step in the geneformer tests I think I did that at one point.

Reply:
The above should make it so you do not need to wrap in a float16module etc.

Reply:
There's also the fabric API, you could ask the NeMo team how close that is to primetime: https://lightning.ai/blog/quickstart-to-lightning-fabric/ that would be the prettiest solution.

Reply:
Yeah I'll have to create a full datamodule for `trainer.predict`, that might be more than I want for now. Let's just keep the Float16Module for now and come back to this when the fabric API has landed.